apps/server/lib/lexical/server/task_queue.ex

```diff
@@ -0,0 +1,201 @@
+    def cancel(%__MODULE__{} = state, request_id) do
+      with {:ok, %Task{} = task} <- Map.fetch(state.ids_to_tasks, request_id),
+           :ok <- TaskQueue.cancel_task(task) do
+        write_error(request_id, "Request cancelled", :request_cancelled)
+
+        %__MODULE__{
+          state
+          | ids_to_tasks: Map.delete(state.ids_to_tasks, request_id),
+            pids_to_ids: Map.delete(state.pids_to_ids, task.pid)
+        }
+      else
+        _ ->
+          state
+      end
+    end
+
+    def size(%__MODULE__{} = state) do
+      map_size(state.ids_to_tasks)
+    end
+
+    def task_finished(%__MODULE__{} = state, pid, reason) do
+      case Map.pop(state.pids_to_ids, pid) do
+        {nil, _} ->
+          state
+
+        {request_id, new_pids_to_ids} ->
+          maybe_log_task(reason, request_id)
+
+          %__MODULE__{
+            state
+            | pids_to_ids: new_pids_to_ids,
+              ids_to_tasks: Map.delete(state.ids_to_tasks, request_id)
+          }
+      end
+    end
+
+    defp maybe_log_task(:normal, _),
+      do: :ok
+
+    defp maybe_log_task(reason, request_id),
+      do: Logger.warning("Request id #{request_id} failed with reason #{inspect(reason)}")
+
+    defp as_task(request_id, {m, f, a}) do
+      handler = fn ->
+        try do
+          case apply(m, f, a) do
+            :noreply ->
+              {:request_complete, request_id}
+
+            {:reply, reply} ->
+              write_reply(reply)
+
+              {:request_complete, request_id}
+          end
+        rescue
+          e ->
+            exception_string = Exception.format(:error, e, __STACKTRACE__)
+            Logger.error(exception_string)
+            write_error(request_id, exception_string)
+
+            {:request_complete, request_id}
+        end
+      end
+
+      TaskQueue.run_task(handler)
+    end
+
+    defp write_reply(response) do
+      case Convert.to_lsp(response) do
+        {:ok, lsp_response} ->
+          Transport.write(lsp_response)
+
+        error ->
+          error_message = """
+          Failed to convert #{response.__struct__}:
+
+          #{inspect(error, pretty: true)}\
+          """
+
+          Logger.critical("""
+          #{error_message}
+
+          #{inspect(response, pretty: true)}\
+          """)
+
+          write_error(response.id, error_message)
+      end
+    end
+
+    defp write_error(id, message, code \\ :internal_error) do
+      error =
+        ResponseError.new(
+          code: code,
+          message: message
+        )
+
+      Transport.write(%{id: id, error: error})
+    end
+  end
+
+  use GenServer
+
+  def task_supervisor_name do
+    __MODULE__.TaskSupervisor
+  end
+
+  @doc false
+  def run_task(fun) when is_function(fun) do
+    Task.Supervisor.async_nolink(task_supervisor_name(), fun)
+  end
+
+  @doc false
+  def cancel_task(%Task{} = task) do
+    Task.Supervisor.terminate_child(task_supervisor_name(), task.pid)
+  end
```

Comment:
why are run_task and cancel_task these here and not in the state?

Reply:
Good question; they could and should be.